File: components/js-evm-price-oracle/trigger.ts
/**
 * Trigger decoding helpers for WAVS operator@2.7.0
 *
 * Handles both on-chain EVM contract events and CLI (raw) inputs.
 *
 * NOTE: No ethers.js — WAVS stubs wasi:random as `unreachable` for
 * determinism, and ethers.js calls crypto.getRandomValues on import.
 * All ABI encode/decode is done manually.
 */
import { TriggerData } from "./out/interfaces/wavs-types-events.js";

export enum Destination {
  Cli = "Cli",
  Ethereum = "Ethereum",
}

export interface TriggerInfo {
  triggerId: number;
  creator: string;
  data: Uint8Array;
}

// ─── ABI helpers ────────────────────────────────────────────────────────────

/** Read a big-endian uint256 from 32 bytes at offset, return as BigInt */
function readUint256(buf: Uint8Array, offset: number): bigint {
  let v = 0n;
  for (let i = 0; i < 32; i++) {
    v = (v << 8n) | BigInt(buf[offset + i]);
  }
  return v;
}

/** Read a uint64 from the low 8 bytes of a 32-byte word */
function readUint64(buf: Uint8Array, offset: number): number {
  // The value is right-aligned in a 32-byte slot
  let v = 0;
  for (let i = 24; i < 32; i++) {
    v = (v * 256 + buf[offset + i]);
  }
  return v;
}

/** Read a 20-byte Ethereum address from the low 20 bytes of a 32-byte word */
function readAddress(buf: Uint8Array, offset: number): string {
  const bytes = buf.slice(offset + 12, offset + 32);
  return "0x" + Array.from(bytes).map(b => b.toString(16).padStart(2, "0")).join("");
}

/** Read `bytes` from an ABI-encoded dynamic type at the given base offset */
function readBytes(buf: Uint8Array, baseOffset: number, dynOffset: bigint): Uint8Array {
  const start = Number(dynOffset);
  const len = readUint64(buf, baseOffset + start);
  return buf.slice(baseOffset + start + 32, baseOffset + start + 32 + len);
}

/**
 * Decode an ABI-encoded `TriggerInfo { uint64 triggerId; address creator; bytes data; }`.
 * The outer NewTrigger event has `abi.encode(triggerInfo)` as the bytes payload,
 * which is itself wrapped in `abi.encode(bytes)` in the event data.
 */
function decodeTriggerInfo(eventData: Uint8Array): TriggerInfo {
  // The event data is: abi.encode(bytes _triggerInfo)
  // Layout: offset(32) | length(32) | triggerInfoBytes
  // The inner triggerInfoBytes starts at byte 64
  const innerOffset = readUint64(eventData, 0);   // should be 32
  const innerLen = readUint64(eventData, 32);
  const inner = eventData.slice(64, 64 + innerLen);

  // abi.encode(TriggerInfo) adds a 32-byte offset prefix before the tuple
  // because TriggerInfo contains a dynamic field (bytes data).
  // So inner = 0x20(32 bytes) + tuple data
  const tuple = inner.slice(32);

  // Tuple layout (after the 0x20 offset prefix):
  // [0:32]   triggerId  (uint64, right-aligned)
  // [32:64]  creator    (address, right-aligned)
  // [64:96]  offset to data field (= 0x60 = 96)
  // [96:128] length of data
  // [128..]  data bytes (padded to 32-byte boundary)
  const triggerId = readUint64(tuple, 0);
  const creator = readAddress(tuple, 32);
  const dataOffset = readUint64(tuple, 64); // should be 96
  const dataLen = readUint64(tuple, 96);
  const data = tuple.slice(128, 128 + dataLen);

  return { triggerId, creator, data };
}

/**
 * ABI-encode the output for on-chain submission.
 * Matches ITypes.sol: struct DataWithId { uint64 triggerId; bytes data; }
 *
 * Equivalent to alloy's `.abi_encode()` on DataWithId — same as Solidity's
 * `abi.encode(dataWithId)` which `abi.decode(payload, (DataWithId))` expects.
 *
 * DataWithId is a dynamic type (contains `bytes`), so the encoding includes
 * a 32-byte outer offset pointer before the struct fields:
 *
 *   [0:32]   = 0x20 (outer offset → struct data starts at byte 32)
 *   [32:64]  = triggerId (uint64, right-aligned)
 *   [64:96]  = 0x40 (offset to bytes, relative to byte 32 → bytes at byte 96)
 *   [96:128] = dataLen
 *   [128:…]  = data (zero-padded to 32-byte boundary)
 */
export function encodeOutput(triggerId: number, outputData: Uint8Array): Uint8Array {
  const dataLen = outputData.length;
  const padLen = (32 - (dataLen % 32)) % 32;
  const total = 32 + 96 + dataLen + padLen;  // +32 for outer offset word

  const buf = new Uint8Array(total);
  const view = new DataView(buf.buffer);

  // Word 0: outer offset — struct data starts at byte 32
  view.setBigUint64(24, 32n, false);

  // Word 1 (byte 32): triggerId — uint64 right-aligned
  view.setBigUint64(56, BigInt(triggerId), false);

  // Word 2 (byte 64): offset to bytes data = 64 (relative to byte 32, so bytes at byte 96)
  view.setBigUint64(88, 64n, false);

  // Word 3 (byte 96): length of bytes data
  view.setBigUint64(120, BigInt(dataLen), false);

  // Byte 128+: bytes data
  buf.set(outputData, 128);

  return buf;
}

/**
 * Decode a WAVS trigger event (on-chain or CLI).
 */
export function decodeTriggerEvent(
  triggerData: TriggerData
): [TriggerInfo, Destination] {
  if (triggerData.tag === "raw") {
    return [
      { triggerId: 0, creator: "", data: triggerData.val },
      Destination.Cli,
    ];
  }

  if (triggerData.tag === "evm-contract-event") {
    const { log } = triggerData.val;

    try {
      // In wavs:operator@2.7.0, log.data is EvmEventLogData { topics: Uint8Array[], data: Uint8Array }
      const eventData = log.data.data;
      const triggerInfo = decodeTriggerInfo(eventData);

      return [triggerInfo, Destination.Ethereum];
    } catch (error) {
      throw new Error("Error processing evm-contract-event: " + error);
    }
  }

  throw new Error(
    "Unsupported trigger data type: " + (triggerData as any).tag
  );
}

File: components/python-evm-price-oracle/trigger.py
"""
ABI decode/encode helpers for the WAVS EVM price oracle.

Manually decodes the NewTrigger event and encodes the DataWithId output —
no external libraries required, pure Python byte manipulation.

Logic ported directly from the Go implementation (trigger.go).
"""

import struct
from typing import Tuple

from wit_world.imports import events


# ─── Destination ─────────────────────────────────────────────────────────────

DEST_CLI = "cli"
DEST_EVM = "evm"


def decode_trigger(data: events.TriggerData) -> Tuple[int, bytes, str]:
    """Decode a WAVS TriggerData into (trigger_id, input_bytes, destination).

    Raw trigger  → CLI path (input_bytes = raw bytes, trigger_id = 0)
    EVM trigger  → on-chain path (input_bytes = CMC ID as UTF-8)
    """
    if isinstance(data, events.TriggerData_Raw):
        raw = data.value.rstrip(b"\x00")
        return 0, raw, DEST_CLI

    if isinstance(data, events.TriggerData_EvmContractEvent):
        log = data.value.log
        # log.data.data = raw ABI bytes from NewTrigger(bytes _triggerInfo)
        trigger_id, input_bytes = _decode_trigger_info(log.data.data)
        return trigger_id, input_bytes, DEST_EVM

    raise ValueError(f"Unsupported trigger type: {type(data).__name__}")


def encode_output(trigger_id: int, data: bytes) -> bytes:
    """ABI-encode DataWithId { uint64 triggerId; bytes data } for on-chain submission.

    Matches Rust alloy's .abi_encode() for a dynamic struct — includes the outer
    32-byte tuple offset prefix that abi.decode(payload, (DataWithId)) requires.

    Layout:
      [0:32]    outer offset = 0x20 (= 32, pointing to struct start)
      [32:64]   triggerId  (uint64, right-aligned in 32 bytes)
      [64:96]   offset to data bytes from struct start (= 0x40 = 64)
      [96:128]  length of data bytes
      [128..]   data bytes (zero-padded to 32-byte boundary)

    Note: abi.decode(payload, (DataWithId)) expects this format —
    it decodes a 1-tuple (DataWithId,) which adds an outer offset word.
    Without this prefix, the decode silently produces wrong values and reverts.
    """
    data_len = len(data)
    pad_len = (32 - data_len % 32) % 32 if data_len % 32 != 0 else 0
    buf = bytearray(128 + data_len + pad_len)

    # outer offset = 32 (0x20) — required by abi.decode for dynamic structs
    struct.pack_into(">Q", buf, 24, 32)

    # triggerId — uint64 right-aligned in second 32-byte word
    struct.pack_into(">Q", buf, 56, trigger_id)

    # offset to data from struct start = 64 (0x40)
    struct.pack_into(">Q", buf, 88, 64)

    # data length
    struct.pack_into(">Q", buf, 120, data_len)

    # data bytes
    buf[128:128 + data_len] = data

    return bytes(buf)


# ─── Internal ─────────────────────────────────────────────────────────────────

def _decode_trigger_info(raw: bytes) -> Tuple[int, bytes]:
    """Decode a NewTrigger event payload into (trigger_id, data_bytes).

    The Solidity event is:
        event NewTrigger(bytes _triggerInfo)
    where _triggerInfo = abi.encode(TriggerInfo{triggerId, creator, data}).

    Outer event ABI layout (bytes as a dynamic type):
      [0:32]   offset to _triggerInfo (= 0x20)
      [32:64]  length of _triggerInfo bytes
      [64:N]   _triggerInfo ABI bytes

    TriggerInfo ABI layout (after skipping 32-byte ABI-prefix for struct):
      [0:32]    triggerId  (uint64, right-aligned)
      [32:64]   creator    (address, right-aligned)
      [64:96]   offset to data bytes (= 0x60)
      [96:128]  length of data bytes
      [128..]   data bytes (CMC ID as UTF-8, padded to 32-byte boundary)
    """
    if len(raw) < 96:
        raise ValueError(f"Event data too short: {len(raw)} bytes")

    # Extract inner length + bytes from the outer bytes wrapper
    inner_len = struct.unpack_from(">Q", raw, 56)[0]   # bytes 56–63 of 32-byte length word
    inner_start = 64
    if len(raw) < inner_start + inner_len:
        raise ValueError(f"Inner data truncated: need {inner_start + inner_len}, have {len(raw)}")

    inner = raw[inner_start: inner_start + inner_len]

    # Skip the 32-byte ABI tuple offset prefix (= 0x20 = 32)
    if len(inner) < 32:
        raise ValueError(f"Inner too short for ABI offset header: {len(inner)} bytes")
    inner = inner[32:]

    # Parse TriggerInfo fields
    if len(inner) < 128:
        raise ValueError(f"TriggerInfo too short: {len(inner)} bytes")

    trigger_id = struct.unpack_from(">Q", inner, 24)[0]   # low 8B of 32-byte word
    # creator = inner[44:64] — not needed for price oracle

    data_len = struct.unpack_from(">Q", inner, 120)[0]    # low 8B of 32-byte length word
    data_start = 128
    if len(inner) < data_start + data_len:
        raise ValueError(f"Trigger data truncated: need {data_start + data_len}, have {len(inner)}")

    data = inner[data_start: data_start + data_len]
    return trigger_id, data
